### PR TITLE
[NO GBP] Loosens gas pressure transfer calculation limit a bit.

### PR DESCRIPTION
**Pull request by vincentiusvin**
## About The Pull Request
I initially kept it an exclusive check so funny things like division by inf doesnt get in, but I implemented more inf checks on the quadratic proc and the likes so it really shouldnt be necessary now. This should hopefully reduce cases when the target pressure hits the upper limit because the target gasmix has a miniscule amount of moles, having the result be rejected by the code, and triggering a runtime.

And some minor improvements to the code so it doesnt iterate through the gaslist multiple times.

## Why It's Good For The Game
Nicer code

## Changelog
:cl:
code: changed how the gas pressure calculation work very slightly, report if things are bugged.
/:cl:

### CODE REVIEW

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 1]
```diff
@@ -578,15 +578,21 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
  * - ignore_temperature. Returns a cheaper form of gas calculation, useful if the temperature difference between the two gasmixes is low or nonexistant.
  */
 /datum/gas_mixture/proc/gas_pressure_calculate(datum/gas_mixture/output_air, target_pressure, ignore_temperature = FALSE)
-	if((total_moles() <= 0) || (temperature <= 0))
+	// So we dont need to iterate the gaslist multiple times.
+	var/our_moles = total_moles()
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm:582]: this value won't change midway through this proc right?

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 1 above)
Comment (vincentiusvin) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm:582]: Yes it shouldnt.

**Comment by vincentiusvin** [comment]
Hmm it messes with null value, on minute

**Review by LemonInTheDark — APPROVED** [review]

**Comment by LemonInTheDark** [comment]
shit no go back

**Comment by vincentiusvin** [comment]
Oh yeah, one thing I found while testing this, the limits i explained in https://github.com/tgstation/tgstation/pull/62320 will be hit sooner if your pressure target is very high. Will also mess with the newton's method because you're bumping the pvr real high. Will never happen in game though.

**Comment by vincentiusvin** [comment]
Hello. The commit I just pushed eb5def0df506faddfa36c24328282fcffa338ccf illustrates our problem here. One of the testcases failed due to a rounding and order of operation issue. 
Basically our lower limit is supposed to be 500 - 0.0001, but this gets skewed to 500.00003 - 0.0001. Making the lower limit lower than the actual real lower limit, failing the test.
image

Simple fix, just needs one source of truth.

**Comment by vincentiusvin** [comment]
Nevermind not a simple fix. Gotta implement a boundary man.

**Comment by vincentiusvin** [comment]
Good to go btw.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 2]
```diff
@@ -659,8 +666,8 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 /datum/gas_mixture/proc/gas_pressure_quadratic(a, b, c, lower_limit, upper_limit)
 	var/solution
 	if(!IS_INF_OR_NAN(a) && !IS_INF_OR_NAN(b) && !IS_INF_OR_NAN(c))
-		solution = max(SolveQuadratic(a, b, c))
-		if((solution > lower_limit) && (solution < upper_limit)) //SolveQuadratic can return nulls so be careful here
+		solution = max(SolveQuadratic(a, b, c)) 
+		if((solution > lower_limit) && (solution < upper_limit)) //SolveQuadratic can return empty lists so be careful here
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm:670]: why the ()s here?

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 3]
```diff
@@ -578,15 +578,20 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
...
+	var/our_moles = total_moles()
+	var/output_moles = output_air.total_moles()
+	var/output_pressure = output_air.return_pressure()
+
+	if((our_moles <= 0) || (temperature <= 0))
 		return FALSE
 
 	var/pressure_delta = 0
-	if((output_air.temperature <= 0) || (output_air.total_moles() <= 0))
+	if((output_air.temperature <= 0) || (output_moles <= 0))
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm:590]: and here

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 4]
```diff
@@ -578,15 +578,20 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
  * - ignore_temperature. Returns a cheaper form of gas calculation, useful if the temperature difference between the two gasmixes is low or nonexistant.
  */
 /datum/gas_mixture/proc/gas_pressure_calculate(datum/gas_mixture/output_air, target_pressure, ignore_temperature = FALSE)
-	if((total_moles() <= 0) || (temperature <= 0))
+	// So we dont need to iterate the gaslist multiple times.
+	var/our_moles = total_moles()
+	var/output_moles = output_air.total_moles()
+	var/output_pressure = output_air.return_pressure()
+
+	if((our_moles <= 0) || (temperature <= 0))
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm:586]: and here now that I think about it

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 2 above)
Comment (vincentiusvin) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm:670]: readability i guess?

Comment (vincentiusvin) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: d4ed90dc9033ec1a732c4042cdc380ffc0b70ac0 should remove those parentheses. I personally like it more the old way, helps me read things quicker while making this, but don't really mind either.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 5]
```diff
@@ -670,8 +677,9 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 /datum/gas_mixture/proc/gas_pressure_approximate(a, b, c, lower_limit, upper_limit)
 	var/solution
 	if(!IS_INF_OR_NAN(a) && !IS_INF_OR_NAN(b) && !IS_INF_OR_NAN(c))
-		/// We need to start off at a reasonably good estimate. For very big numbers the amount of moles is most likely small so better start with lower_limit.
-		solution = lower_limit
+		// We start at the extrema of the equation, added by a number.
+		// This way we will hopefully always converge on the positive root, while starting at a reasonable number.
+		solution = (-b / (2*a)) + 200
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm:682]: spaces between operators?

**Review by LemonInTheDark — APPROVED** [review]
outside of the above this looks good. I trust you

**Pull request merged**